Lecture4/Tasks1-3/Task1_3L4.java

```diff
@@ -0,0 +1,318 @@
+                selectVoucher();
+            } else {
+                searchMinimumDaysCount();
+                searchWithFood();
+                searchWithTransfer();
+                provideFinalOffer();
+            }
+        } catch (OutOfMemoryError | NullPointerException mainFlowEx) {
+            System.out.println(MESSAGE_OPS);
+            System.exit(0);
+        }
+    }
+
+    private static void readVouchersFromFile() {
+        int notDownloadedCount = 0;
+        try {
+            BufferedReader readVouchersFile = new BufferedReader(new FileReader("initVouchers.txt"));
+            String sCurrentLine;
+            while ((sCurrentLine = readVouchersFile.readLine()) != null) {
+                String[] voucherDetails = sCurrentLine.split(";");
+                try {
+                    switch (Integer.parseInt(voucherDetails[0])) {
+                        case 1:
+                            Voucher restVoucher = new RestVoucher(voucherDetails);
+                            listOfVouchers.add(restVoucher);
+                            break;
+                        case 2:
+                            Voucher cureVoucher = new CureVoucher(voucherDetails);
+                            listOfVouchers.add(cureVoucher);
+                            break;
+                        case 3:
+                            Voucher shoppingVoucher = new ShoppingVoucher(voucherDetails);
+                            listOfVouchers.add(shoppingVoucher);
+                            break;
+                        default:
+                            notDownloadedCount++;
+                    }
+                } catch (NumberFormatException
+                        | ArithmeticException
+                        | ArrayIndexOutOfBoundsException
+                        | RegexException initEx) {
+                    notDownloadedCount++;
+                }
+            }
+        } catch (IOException inputEx) {
+            inputEx.printStackTrace();
+            System.exit(0);
+        }
+        if (notDownloadedCount != 0) {
+            System.out.println("\nThere are some problems with " + notDownloadedCount + " vouchers within the file\n");
+        }
+    }
+
+    private static int getUserSelection() {
+        int selection = 0;
+        while (selection == 0) {
+            System.out.print("Lets select needed type of the vouchers (1-Rest, 2-Cure, 3-Shopping): ");
+            Scanner input = new Scanner(System.in);
+            int userInput;
+            try {
+                userInput = input.nextInt();
+            } catch (InputMismatchException inputMismatchEx) {
+                System.out.print(MESSAGE_TRY_AGAIN);
+                continue;
+            }
+            try {
+                if (userInput >= 1 && userInput <= 3) {
+                    selection = userInput;
+                } else {
+                    throw new SelectedVoucherTypeException(MESSAGE_TRY_AGAIN);
```

Comment:
throwing and catching the same exception make no sense in the same code block

Reply:
fixed in Tasks1-4-fixed package